### DISCUSSION TITLE
Early feedback

### DISCUSSION THREAD

**Discussion by VojtechVitek** [golang-cz/go2webrpc#2] [General]
Please help us shape gospeak@v1 by providing an early feedback.

**Comment by ftjfo** [comment]
I just want to add some encouraging words to your project.
As someone who is using WebRPC and is happy with it!  I would rank this "code-first" approach as an even better, at least for my use cases with Go<->Typescript RPC.  I am surprised that this way of creating interfaces doesn't have more traction in general.

**Comment by gsouf** [comment]
Hey, just found out about it. Had hard time deciding between webrpc and connect-rpc. I was into webrpc more for the ease of creating new clients,  but what bothers me most was the need to use a DSL (ridl) to write my routes.

This one might be the way to go actually  ("go" pun unintended)

**Reply by gsouf** [reply]
@VojtechVitek  I was able to play quickly with it. A few things caught my attention.


### 1) generation creates a new struct
 
Firstly, defining a custom type for the generation, for instance the Tag from the examples in the README:

```go
//go:webrpc golang -server -pkg=server -out=./server/server.gen.go
type PetStore interface {
   ...
}

type Tag struct {
	ID   int64
	Name string
}
```
When it's generated in the server, it's generating a new Tag struct in `server.gen.go`, instead of reusing that one. It would be ideal if it could just reuse the same struct for consistency across the code base and better integration with IDE tools

### 2) Cannot use new types
 I was hoping that using go rather than ridl would make it possible to use new types. For instance:

```go
type UUID string

type Tag struct {
	ID   int64
	Name UUID  // <====== note the UUID here
}
```

In the generated server, it used the same UUID type:

```
type Tag struct {
		ID UUID `json:"ID"`
		Name string `json:"Name"`
}
```

but UUID is undefined. It's somehow related to my previous observation that it creates a new struct, without importing the new type.

### 3) It does not pass struct tags

I used a custom struct tag to integrate with validate:

```go
type Tag struct {
	ID   int64
	Name string `validate:"required,max=100"`
}
``` 

but in the generated code it's just discarded.

Being able to do that would help to do many things with middlewares in an expressive way.

-------

Are there limitation inherent to the fact it's backed by webrpc? Or are there things we might see implemented in gospeak?

**Reply by VojtechVitek** [reply]
Hi @gsouf, please try to use `-types=false` flag and see if it helps with the above.

```
//go:webrpc golang -server -types=false -pkg=proto -json=sonic -errorStackTrace -out=server.gen.go
```

**Reply by gsouf** [reply]
@VojtechVitek  almost! It's not recreating the types, but also the types don't get imported in the generated server file. Therefore I  get errors such as `undefined : PetStore`, `undefined: Pet`

**Reply by gsouf** [reply]
Sorry my bad, I had to adjust the `out` param too, to make sure they land in the same directory. That looks great now!

**Reply by gsouf** [reply]
One thing to note, setting types to false it now leaves some empty placeholder for types:

```
var WebRPCServices = map[string][]string{
	"PetStore": {
		"CreatePet",
		"DeletePet",
		"GetPet",
		"ListPets",
		"UpdatePet",
	},
}

//
// Server types
//








//
// Client types
//



//
// Server
//

type WebRPCServer interface {
	http.Handler
}
```

Other than that it looks really awesome! I see great potential with expressiveness and simplicity. That is something that I truly value!

Great job ✨ 🚀 

I'll start working on some concrete implementation tomorrow, I haven't been so excited to write types in many years 🥲

**Reply by VojtechVitek** [reply]
Great! Thanks for any feedback along the way 👍

**Reply by gsouf** [reply]
@VojtechVitek is there a way to define the possible errors that the a given endpoint can return? 

I haven't tried webrpc itself but I see that the RIDL format as a spec to define possible global errors; i.e:

```
error 100 RateLimited     "too many requests"   HTTP 429
error 101 DatabaseDown    "service outage"      HTTP 503
```

But I haven't seen that with gospeak. Also I haven't seen specifically how to attach a specific type of a error that a given endpoint can return.

The goal is to make it possible to use type safe and exhaustive error matching in the response returned by the client.

Is that achievable today?

**Reply by VojtechVitek** [reply]
This is not achievable yet. It's on my TODO.

Currently, I'm doing this on the server side:
```
// type WebRPCError = gospeak.WebRPCError

var (
	// General.
	ErrTODO           = WebRPCError{Code: 1000, Name: "NotImplemented", Message: "TODO: Not implemented", HTTPStatus: 500}
	ErrDeprecated     = WebRPCError{Code: 1001, Name: "Deprecated", Message: "Endpoint is deprecated", HTTPStatus: 410}
	ErrRateLimited    = WebRPCError{Code: 1002, Name: "RateLimited", Message: "Rate limited. Please, slow down", HTTPStatus: 429}
```

It partially works.. But has some go:generate issues.. and the semantics of the errors doesn't propagate to the client-generated code yet. So it's not 100% ready yet.

**Reply by gsouf** [reply]
@VojtechVitek Is there anyway I can help getting there?

**Reply by VojtechVitek** [reply]
What's currently missing in gospeak vs. webrpc:

Design:
- Error definition
- ENUM definition v2 -- a prototype is at https://github.com/golang-cz/gospeak/tree/enum_v2
- Webrpc streaming

We need to come up with the right design first.

Gospeak is already usable right now -- but it's not there 100%. There are missing features & rough edges with `go generate`. We need to provide more examples and better best practices, e.g. should interface definition be side-by-side to the generated files or in a separate folder?

I don't have much time to improve any of this right now. But I will revisit during the XMas break.

I'm all ears for design ideas!

**Reply by gsouf** [reply]
@VojtechVitek  I strongly believe in this project and I'd be happy to contribute to make it better.

Regarding error management specifically. My need today for a better typed client is to be able to identity which error can the endpoint return. For example in extremely simplified pseudo code:

```
response = client.getPets();

// Use an **Exhaustive** switch statement
switch response {
  authError:
    // do something
  validationError:
    // do something
  customPerError:
    // do something
  otherCustomPetError:
     // do something
  noError:
      // do something
}
```


Because the switch here is exhaustive, whenever a new error is added, this switch statement will no longer be valid and we will have to review it. 

Also it's worth it to add that some errors might come from generic middlewares, other errors will be specific to one or more endpoints, related to business logic. 

I'm wondering whether it would be beneficial to make it possible to distinct them somehow for the client side.

I'm very new to the RPC/webRPC standard so I might also have overseen something here.. My goal being to make it really easy for the client to detect any change from the client via strong typing.

**Reply by VojtechVitek** [reply]
The errors are currently global in webrpc: https://github.com/webrpc/webrpc/blob/master/_examples/hello-webrpc-ts/webapp/src/client.gen.ts#L256

So, you'd have to handle all possible error types with the exhaustive switch statement.

However, we discussed something similar recently -- and there's a "throw" proposal at https://github.com/webrpc/webrpc/pull/383 to make the errors more fine-grained for each endpoint. It's a very very early concept, though.

**Reply by gsouf** [reply]
@VojtechVitek is it possible to add more than one service in the same generated client / server file?

**Reply by VojtechVitek** [reply]
Good question. I've tested this not a long time ago and there were some bugs when using multiple interfaces.

So I the answer is no as of today. But I'd love to revisit this. The end goal is to be able to define multiple services as separate interfaces within one file:

```
type PublicAPI interface {
     GetUser(context.Context)
}

type AdminAPI interface {
     CreateUser(context.Context)
}
```

**Reply by gsouf** [reply]
@VojtechVitek  I do have another question.

I have an endpoint that returns data of unknown format. I'd like to return any nested object here .  I haven't found a good solution for that. At the moment I simply json encode my data and put it in a string, but that's not efficient because it's getting json encoded twice and the caller using the js client has to manually json decode the value to use it. 

To illustrate, my handler returns this struct:


```
type SomeResponse struct {
	JsonEncodedString string `json:"data"`   
}
```

If you wonder about the use case, it's data returned by a third party api for which we have no control on the format returned ; and the format depends on the input.


Is there any way to do that better with gospeak?

**Reply by VojtechVitek** [reply]
Have you tried type `any`?

**Reply by gsouf** [reply]
@VojtechVitek thanks for the suggestion, when I use `any` the data gets base64 encoded. I found that in the generated code of the golang generator, it simply uses json.Marshal to encode the data, therefore I had to use `json.RawMessage` to get it working as I expected:

```
type SomeResponse struct {
	Data json.RawMessage `json:"data"`
}
```

That's the thing I love with web rpc and gospeak, it's really easy to debug things as the code generated is not crazy complex or using deep abstraction 🚀
